src/uct/self/self_ep.c

```diff
@@ -23,19 +23,81 @@ static UCS_CLASS_CLEANUP_FUNC(uct_self_ep_t)
     ucs_trace_func("self=%p", self);
 }
 
-UCS_CLASS_DEFINE(uct_self_ep_t, uct_base_ep_t)
+UCS_CLASS_DEFINE(uct_self_ep_t, uct_base_ep_t);
 UCS_CLASS_DEFINE_NEW_FUNC(uct_self_ep_t, uct_ep_t, uct_iface_t *,
                           const uct_device_addr_t *, const uct_iface_addr_t *);
 UCS_CLASS_DEFINE_DELETE_FUNC(uct_self_ep_t, uct_ep_t);
 
+static void
+uct_self_ep_am_reserve_and_get_new_buffer(ucs_status_t *status,
+                                          uct_self_iface_t *self_iface,
+                                          const uct_self_msg_desc_t *desc)
+{
+    /* Reserve the buffer */
+    uct_recv_desc_iface(desc->am_recv_data) = &self_iface->super.super;
+
+    /* get new buffer from memory pool
+    * No need to copy data from payload to desc->am_recv_data */
+    self_iface->msg_cur_desc = ucs_mpool_get(&self_iface->msg_desc_mp);
+    VALGRIND_MAKE_MEM_DEFINED(self_iface->msg_cur_desc,
+                              sizeof(*(self_iface->msg_cur_desc)));
+    if (NULL == self_iface->msg_cur_desc) {
+        ucs_error("Failed to get next descriptor in SELF MP storage");
+        *status = UCS_ERR_NO_RESOURCE;
```

Comment:
@yosefe The function uct_self_ep_am_reserve_and_get_new_buffer called if return status != UCS_OK (it should be UCS_INPROGRESS). In this case I have to keep the current buffer (reserve it for the user later release by uct_self_iface_release_am_desc) and get the new one from MP.
In this case the message hasn't arrived to the user buffer and (if no memory) NO_RESOURCE will be returned.
Should I implement some procedure (before uct_iface_invoke_am) to** check if new buffer can be obtained later (after uct_iface_invoke_am called)?